### ISSUE TITLE
H2O mod_rewrite support

### ISSUE THREAD

**Issue by kerbenii** [h2o/h2o#678]
Is there any chance to support original .htaccess mod_rewrite?

I would like to battle test H2O on shared webhosting environment,
but almost all users use RewriteEngine.

**Comment by webhtb** [comment]
Use mruby, works fine, here is an example:

hosts:
  "host.com:80":
    listen:
      host: ......
      port: 80
    paths:
      "/":
        mruby.handler: |
          lambda do |env|
            if /rewrite\/([^\/]+)/.match(env["PATH_INFO"])
              return [307, {"x-reproxy-url" => "http://host.com/rewrite.php?q=#{$1}"}, []]
            end
            return [399, {}, []]
          end
        reproxy: ON
        file.dir: .....

**Comment by kerbenii** [comment]
Yes, I know about mruby.handler, but it is not enough for shared webhosting environment.
If I change webserv to H2O, almost all clients websites crash.
And, on the other hand I can not force all users to recreate their RewriteEngine to mruby.

Due to above, please consider supporting original mod_rewrite syntax.
This will definitely have positive impact on the popularity of  H2O server.

**Comment by webhtb** [comment]
Yes, also my H2O crash when i make more rewrite rules, in 10 seconds, sometime 1 minute, the h2o just die. I use h2o only for my projects, i cant force the users to change the confs.

**Comment by utrenkner** [comment]
@webhtb I don't think we want to have h2o as 1:1 replacement for Apache. 

There is a reason why no other web server offers this: When you allow the user to configure rewrites via .htaccess files the webserver has to recursively look for these files in each folder in the path to the requested resource - for each and every single request!

Just look at a single stylesheet in a typical Wordpress installation:
https://mydomain.com/wp-content/plugins/contact-form-7/includes/css/styles.css

In this case, Apache has to look for .htaccess in
/
/wp-content/
/wp-content/plugins/
/wp-content/plugins/contact-form-7/
/wp-content/plugins/contact-form-7/includes/

5 places to check for possible .htaccess files!!! And this is only for one resource - a typical WP site may request 30 or even more resources...

Rewrites are ok. But doing this via .htaccess files is creating too much overhead in my opinion. Let's not go down this route. Those who need .htaccess should stick with Apache.

**Comment by kerbenii** [comment]
Yes, but we can add this as additional module 
(or create parser which will parse on request all   .Htaccess files to one  mruby).

**Comment by webhtb** [comment]
i not requested  this, i just tell when i create some rewrite rules with mruby, the h2o just die, soon or later, sometime from start

**Comment by kazuho** [comment]
@kerbenii @webhtb
If you are looking for a rewrite engine that is compatible with mod_rewrite, I would suggest implementing it in mruby and sharing it with others.

In addition to the things pointed out by @utrenkner, we fear that trying to provide compatibility to a large and complicated software such as Apache will drag down the development of H2O as a whole.  Therefore, we are backwards against providing such feature as part of H2O core.

@webhtb 
> i just tell when i create some rewrite rules with mruby, the h2o just die, soon or later, sometime from start

If that is your case, please open a dedicated issue for the topic and provide the mruby code that causes the crash.  It would be great if you could provide a minimal set of configuration and test procedure to reproduce the issue, but that is not always a requirement.

**Comment by mytbk** [comment]
I haven't used the htaccess style rewrite before. I'm now using Caddy to host a mybb forum, which uses rewrite to make a good-looking URL, and the Caddy rewrite works well.
I don't know how much performance cost will it be to implement rewrite with mruby instead of a native rewrite. Also, I think server side rewrite and a 307 redirect are different. I don't know if mruby can do a server side rewrite.

**Comment by ksaltik** [comment]
Yes apache look htaccess file for every folder it finds. what h2o web server is doing It runs every time mruby file.

As @webhtb states "Yes, also my H2O crash when i make more rewrite rules, in 10 seconds, sometime 1 minute, the h2o just die."

h2o server is not performant it just lack of functionality. This is it.

And the developer is stubborn and don't want to listen anyone's suggestion and do it.

Play with your h2o server by yourself because no one will use it unless you give up such behavior.

**Comment by utrenkner** [comment]
@ksaltik You are wrong in many ways. Most people here, find H2O to be very performant. Just because it does not work for you does not mean it is bad or not performant.

Yes, H2O reads mruby files every time. But while Apache tries to find and then read each .htacces file at every request, H2O reads mruby files every time the server is started/reloaded. Big difference. In that respect mruby files in H2O are more like static configuration files in other web browsers (e.g. Apache's httpd.conf).

And just because the developers do not implement the features you would like to see, I would not call them stubborn. Look at who they work for, who pays their trips to HTTP/3 standardization meetings etc. etc. Maybe then you will better understand their priorities.

"no one will use it" is kind of funny, when you consider that most of the work on H2O nowadays is paid for by Fastly, one of the major CDNs on the web. They already use it to deliver web content from thousands of paying customers.

I just came back from the performance.now() conference, where one of the top experts on web performance (webpagetest.org's Pat Meenan) credited H2O as the "gold standard" when it comes to HTTP/2 prioritization. Not bad, if you ask me.

But hey, if you are not happy with the development: The code is MIT licensed. You are free to fork it into any direction you like.

(NOTE: I do not work for Fastly and I am not a H2O developer... just a happy user)
